pytest: Fix shell tests on IOTDK and HSDK

M-Moawad opened the pull request:
This change adds a board-level `testing.flash_before flag` in Twister so the pytest/shell harness flashes the board before opening the serial port (enabled on snps/hsdk, snps/hsdk4xd, and snps/iotdk). It also introduces a configurable `CONFIG_ARC_IOT_CUSTOM_CPU_IDLE` option for the arc_iot SoC that skips sleep to keep peripherals powered when needed.

Why:

Avoid USB CDC disconnects during flashing on ARC development boards
Prevent UART power loss on iotdk when entering sleep

Fixes: #99911

M-Moawad committed:
soc: arc_iot: add configurable custom CPU idle for UART retention

The ARC IoT SoC (used on iotdk board) has a hardware limitation where
entering CPU sleep mode causes some peripherals (e.g., UART) to lose
power, preventing the board from waking up via peripheral interrupts.

Add a configurable custom CPU idle option (CONFIG_ARC_IOT_CUSTOM_CPU_IDLE)
that enables interrupts without executing the sleep instruction. This
allows the system to remain responsive while keeping peripherals powered.

The option defaults to disabled (n) to preserve normal CPU sleep behavior
for most use cases. Tests that require continuous UART operation during
idle (pytest/shell harnesses) can enable it via platform-specific
extra_configs.

This approach allows timer-based idle to work correctly by default while
providing a workaround for tests that need UART retention.

Signed-off-by: Mohamed Moawad <moawad@synopsys.com>

M-Moawad commented on scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py:
```diff
@@ -148,6 +148,11 @@ def _flash_and_run(self) -> None:
+                if self._serial_connection and self._serial_connection.is_open:
+                    self._serial_connection.close()
+                    time.sleep(0.5)
+                    self._connect_device()
```
Thank you for your feedback!

The `--flash-before` flag fixes the two shell tests on boards but causes multiple other tests to fail because it skips reading early boot messages that many tests depend on.
 
 Should we run these two tests separately from the full regression suite, or is there a better approach?
 
My reconnection patch addresses the stale serial connection issue without changing test semantics . Do you think this approach is preferable to using `--flash-before` globally?

M-Moawad replied:
@gchwier 
Yes, `--flash-before` is required for all specified test scenarios for ARC targets (hsdk, hsdk/arc_hsdk/2cores, hsdk4xd, and iotdk) because they fail on CI without it. 

For `vt100_colors_off`, I confirmed it’s specifically needed for `iotdk`; removing the flag causes the prompt wait to fail after flashing:
```bash

============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-8.4.2, pluggy-1.6.0 -- /opt/python/venv/bin/python3
cachedir: .pytest_cache
rootdir: /verification_workspace
collecting ... collected 1 item
zephyr/scripts/pylib/shell-twister-harness/test_shell.py::test_shell_harness
-------------------------------- live log setup --------------------------------
DEBUG: Get device type "hardware"
DEBUG: Opening serial connection for /dev/serial-...
DEBUG: rff: resetting device #5 (SN ...)... OK
DEBUG: Flashing command: /opt/python/venv/bin/west flash --no-rebuild --build-dir /verification_workspace/twister-out/iotdk_arc_iot/zephyr/samples/subsys/testsuite/pytest/shell/sample.harness.shell.vt100_colors_off -- --serial=...
DEBUG: Flashing finished
DEBUG: Found matching key: CONFIG_SHELL_PROMPT_UART="uart:~$ "
INFO: Wait for prompt
ERROR
------------------------------ live log teardown -------------------------------
DEBUG: Closed serial connection for /dev/serial-...
```

gchwier replied:
You fixed it only for those couple of pytest scenarios. What about other tests where ARC target can be used?
I see that `flash_before` is also used for non-pytest harnesses, when `serial_pty` is used (see https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/handlers.py#L783).
I have now also concerns to use `pytest_args: [--flash-before=true]`. If you run it with `--flash-before` in args, or with `flash_before` in hardware map, then you will send it to pytest twice.

You can try to update method `get_hardware`:
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/handlers.py#L703
and add befeore return:
```
        if self.instance.platform.arch == "arc":
            hardware.flash_before = True
```
I do not know if it is the best place to update `flash_before` parameter, but there you can read `platform.arch`. And with that change you do not need to update any test ... flash_before will be set for every ARC target, not needed to add any args.

ruuddw replied:
Synopsys also has simulator boards (nSIM), I'd still prefer some board-level solution. Future boards may handle jtag reset differently as well.

M-Moawad committed:
twister: Add board-level flash_before support for pytest/shell harnesses

Add support for boards to specify flash_before in their board YAML files,
which is then used by pytest and shell harnesses. This is needed for
Synopsys ARC development boards (hsdk, hsdk/arc_hsdk/2cores, hsdk4xd
and iotdk) where the USB serial port disconnects during flashing.

Signed-off-by: Mohamed Moawad <moawad@synopsys.com>

ruuddw reviewed (changes requested):
See comments, please move the custom sleep from arch/arc to soc.

M-Moawad commented on arch/arc/core/cpu_idle.S:
```diff
@@ -43,12 +43,23 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
+ * some peripherals (eg: uart) will power off and board can't
+ * wake up through peripherals interrupt.
+ */
+#ifdef CONFIG_BOARD_IOTDK
```
I see some failed checks here on the CI, and I am checking them

gchwier reviewed (commented):
> To address @gchwier concern "If you run it with --flash-before in args, or with flash_before in hardware map, then you will send it to pytest twice.

My concern was to use `--pytest-args` (but you removed those changes).
Please clean up commits (just remove instead of revert).

gchwier commented on scripts/pylib/twister/twisterlib/handlers.py:
consider also updating docs with `flash_before` parameter:
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/test/twister.rst#L201
https://docs.zephyrproject.org/latest/develop/test/twister.html#board-configuration -> testing

gchwier reviewed (dismissed):
The Twister part looks OK

NIT: however, it would be better to review and test the logic here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/handlers.py#L741
to ensure that `flash_before` is also applied to non‑pytest tests (this could be handled by someone else in a separate PR if there is a need for it)

M-Moawad commented:
Hi @gchwier,

The flash_before commit is unchanged from your previous approval. I've updated the iotdk sleep/idle commit, which removed your approval. Could you please take another look?

M-Moawad commented:
Hi @ruuddw,

 I've updated the iotdk sleep/idle commit. Could you please take another look?

ruuddw reviewed (approved):
Thanks, all looks good to me now. CI seems to be some infrastructure issue ("No space left on device").
